Add operator to column name

Kal-Aster opened the pull request:
Here you can specify the operator in the keys of $conditions array right after the column name, separated by a space character.

It will be converted to the correct operator based on the condition value.

If value is `NULL`:
- `'column ='` => `` `column` IS NULL ``
- `'column !='` => `` `column` IS NOT NULL ``
- `'column <>'` => `` `column` IS NOT NULL ``
- any other operator will throw an `UnexpectedValueException`

If value is an array:
- `'column ='` => `` `column` IN (...) ``
- `'column !='` => `` `column` NOT IN (...) ``
- `'column <>'` => `` `column` NOT IN (...) ``
- any other operator will throw an `UnexpectedValueException`

If anything else it will be kept as is, except for `!=` that will always be converted to `<>`.

The column name will always be trimmed, so it won't ever have trailing or leading spaces.

**Any space character is mandatory between column name and operator**

Kal-Aster committed:
Add support for NULL in fetch and fetchAll $conditions (using: `NULL`, `'NULL'`, `'NOT NULL'`)

Kal-Aster committed:
- Uses $key for operator specification

lulco commented on src/Database/Adapter/PdoAdapter.php:
```diff
@@ -237,14 +238,18 @@ private function createWhere(array $conditions = [], string $where = ''): string
-        $inConditions = [];
-        foreach (array_keys($value) as $index) {
-            $inConditions[] = $this->createValue($key, 'where_' . $index . '_');
+        $rightOperand = null;
```
`$rightOperand` is always set to something, so this line can be removed

lulco commented:
Thank you for your work!
I've added some notes, but it looks very good.

I would appreciate little bit more tests (for each supported operator and also checks for UnexpectedValueException).
Also please fix PHPStan issue(s) - your method has no typehints so please add them.

Then we can merge to master

lulco commented on src/Database/Adapter/PdoAdapter.php:
```diff
@@ -403,6 +409,50 @@ private function isMulti(array $data): bool
+        $operator = '=';
+
+        // check presence of operator in $key
+        if (preg_match('/^(.*)\s(=|!=|<>|<|<=|>|>=)$/', $key, $matches) === 1) {
```
is there any reason to support any white space? isn't one single space ` ` enough?

Kal-Aster replied:
Actually no reason, probably it would be enough

lulco commented on src/Database/Adapter/PdoAdapter.php:
```diff
@@ -403,6 +409,50 @@ private function isMulti(array $data): bool
+        }
+        
+        // always prefer '<>' to '!='
+        if ($operator === '!=') {
```
Is this any sql recommendation?

Kal-Aster replied:
[https://stackoverflow.com/a/13740785/9228492](https://stackoverflow.com/a/13740785/9228492) actually can't find any official documentation

lulco replied:
ok thanks :)

Kal-Aster committed:
- Add typehints
- Match single space instead of any white character
- Add test for UnexpectedValueException

lulco commented on src/Database/Adapter/PdoAdapter.php:
```diff
@@ -237,14 +238,17 @@ private function createWhere(array $conditions = [], string $where = ''): string
-        $inConditions = [];
-        foreach (array_keys($value) as $index) {
-            $inConditions[] = $this->createValue($key, 'where_' . $index . '_');
+        $this->splitColumnNameAndOperator($key, $value, $columnName, $operator);
```
what about to move initialize to this place (before splitColumnNameAndOperatorCall)?
```
// initialize both column name and operator
$columnName = $key;
$operator = '=';
```

Kal-Aster replied:
I put the initialization inside the function, because it is called in both `addCondition` and `bindCondition` so it would be redundant initialize those variables before `splitColumnNameAndOperator` call

Kal-Aster replied:
Well, I see that adding typehint to the arguments, it breaks.

Probably it's better outside, or maybe `splitColumnNameAndOperator` should return `[string, string]` ([ columnName, operator ]) instead of working on referenced variables.

Edit: or use `mixed` or `?string` as typehint

lulco replied:
Yes, I wanted to suggest to you something like that - return array with keys columnName and operator and string values instead of using references

lulco commented on src/Database/Adapter/PdoAdapter.php:
```diff
@@ -403,6 +408,50 @@ private function isMulti(array $data): bool
         return true;
     }
 
+    private function splitColumnNameAndOperator(string $key, $value, string &$columnName, string &$operator): void
```
please add phpdoc typehint mixed for $value

Kal-Aster replied:
$value has no typehint in many places, before my edit. Is it needed everywhere?

lulco replied:
There should be typehints everywhere - at least php doc (for mixed types) because phpstan requires them. I have to check it but I am pretty sure that it is not missing anywhere just here in your new method

Kal-Aster replied:
I saw it missing somewhere, I'll add them

lulco commented on tests/Database/Adapter/PdoAdapterTest.php:
```diff
@@ -121,6 +122,35 @@ public function testFetchAll()
+            $this->assertArrayHasKey('title', $item);
+            $this->assertNull($item['title']);
+        }
+        
```
can you provide also some tests with operators `>` / `<` e.g. id > 2, id < 5 etc?

lulco commented:
Oh, would you be kind too and update also CHANGELOG.md and add some docs about this feature (you can add it here https://github.com/lulco/phoenix/tree/master/docs/migrations)?

Kal-Aster committed:
- improve docs
- improve testFetchAllWithConditions
- fix typehints
- update CHANGELOG

lulco commented on src/Database/Adapter/PdoAdapter.php:
```diff
@@ -403,6 +408,50 @@ private function isMulti(array $data): bool
         return true;
     }
 
+    private function splitColumnNameAndOperator(string $key, mixed $value, ?string &$columnName, ?string &$operator): void
```
mixed typehint is alloed only from php 8. This library supports also PHP >= 7.1 so we need to use just php doc annotation for it like in other cases above

Kal-Aster replied:
Ah! That's why I saw it missing... ahah ok I'll fix that

lulco replied:
yes it can be :)

lulco commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
 ### [Unreleased][unreleased]
 
+### [1.8.1] - 2021-06-01
```
this line will be added when I will release next version :)

lulco commented on docs/migrations/index.md:
````diff
@@ -85,3 +85,39 @@ $this->execute('CREATE TABLE `first_table` (
 );
 ```
+
+### Fetch data
````
👍

lulco commented:
I think we are very close, just finish it :)

Kal-Aster committed:
- return `[ columnName, operator ]` in `splitColumnNameAndOperator`
- fix tests

Kal-Aster committed:
- fix typehint

lulco commented:
Thank you very much, I'll finish this :)

Kal-Aster commented:
Has this been published as new version?

lulco commented:
Hi, it will be released next week. I need to finish something in views support

Kal-Aster commented:
Ok, thanks!

lulco commented:
@Kal-Aster hi, I just released version 1.9.0

Kal-Aster commented:
Thankss!
